sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala

```diff
@@ -206,6 +201,36 @@ class StreamExecution(
     startLatch.await()  // Wait until thread started and QueryStart event has been posted
   }
 
+  private def generateLogicalPlan: LogicalPlan = {
+    var nextSourceId = 0L
+    val internalLogicalPlan = analyzedPlan.transform {
+      case StreamingRelation(dataSource, _, output) =>
+        // Materialize source to avoid creating it in every batch
+        val metadataPath = s"$checkpointRoot/sources/$nextSourceId"
+        val source = dataSource.createSource(metadataPath)
+        nextSourceId += 1
+        // We still need to use the previous `output` instead of `source.schema` as attributes in
+        // "df.logicalPlan" has already used attributes of the previous `output`.
+        StreamingExecutionRelation(source, output)
+    }
+    sources = internalLogicalPlan.collect { case s: StreamingExecutionRelation => s.source }
+    uniqueSources = sources.distinct
+    internalLogicalPlan
+  }
+
+  override def logicalPlan: LogicalPlan = {
+    if (_logicalPlan == null) {
+      localPlanLock.synchronized {
+        if (_logicalPlan == null) {
+          _logicalPlan = generateLogicalPlan
+        }
+      }
+    }
+    _logicalPlan
+  }
```

Comment:
What I'm trying to say is, (if I understand correctly) all of this extra logic is to make sure that there aren't cycles when we fail to initialize and then try to use that uninitialized information in order to generate the exception.

I'm asking why that information needs to be in the exception at all?  Or if we should just try to initialize earlier in the stream thread and throw a different exception when initialization fails.